src/main/java/org/jenkinsci/plugins/workflow/steps/scm/SCMStep.java

```diff
@@ -110,22 +110,26 @@ public final void checkout(Run<?,?> run, FilePath workspace, TaskListener listen
             SCMRevisionState baseline = null;
             Run<?,?> prev = run.getPreviousBuild();
             if (prev != null) {
+                synchronized (prev) {
                 MultiSCMRevisionState state = prev.getAction(MultiSCMRevisionState.class);
                 if (state != null) {
                     baseline = state.get(scm);
                 }
+                }
             }
             scm.checkout(run, launcher, workspace, listener, changelogFile, baseline);
             SCMRevisionState pollingBaseline = null;
             if (poll || changelog) {
                 pollingBaseline = scm.calcRevisionsFromBuild(run, workspace, launcher, listener);
                 if (pollingBaseline != null) {
+                    synchronized (run) {
                     MultiSCMRevisionState state = run.getAction(MultiSCMRevisionState.class);
                     if (state == null) {
                         state = new MultiSCMRevisionState();
                         run.addAction(state);
                     }
                     state.add(scm, pollingBaseline);
```

Comment:
Since this is in a synchronized block now, it shouldn't be possible for the map to be updated while a save is occurring.